### ISSUE TITLE
Add NewType() to create simple unique types with zero runtime overhead

### ISSUE THREAD

**Issue by gvanrossum** [python/typing#189]
The description is best gleaned from the following mypy issue: https://github.com/python/mypy/issues/1284#issuecomment-199021176 and following. We're going with option (A).

Really brief example:
```python
from typing import NewType
UserId = NewType('UserId', int)
```
Now to the type checker UserId is a new type that's compatible with int, but converting an int to a UserId requires a special cast form, UserId(x). At runtime UserId instances are just ints (not a subclass!) and UserId() is a dummy function that just returns its argument.

**Comment by ilevkivskyi** [comment]
I think this is a useful feature.

I have added PR #226  with a simple minded implementation of ``NewType``, some tests, and a short discussion in the PEP.

Also I added ``NewType``, ``Type`` and ``ContextManager`` to the list of things provided by ``typing.py``.

I do not discuss whether ``NewType('EURtoUSD', Callable[[float], float])`` etc. are allowed. If necessary, we could clarify this later.

**Comment by gvanrossum** [comment]
I would like to limit the arg to a class. Also note that that class must have a constructor that is compatible with that assumed by the type checker -- because whenever you write `UserId(42)` the runtime will actually see `int(42)` -- which is fine, and ditto if you use e.g. strings, since `str('x')` is fine, but if you have your own custom class that you want to diversify this way, you have to make sure that your constructor accepts instances of itself -- even though the customary way of constructing instances might be something totally different. This won't work:
```
class PacketId:
    def __init__(self, major: int, minor: int) -> None:
        self._major = major
        self._minor = minor
    # Etc.

TcpPacketId = NewType('TcpPacketId', PacketId)

a = TcpPacketId(127, 0)
x = PacketId(100, 100)
b = TcpPacketId(x)  # ???
```

**Comment by ilevkivskyi** [comment]
> I would like to limit the arg to a class.

OK

> This won't work: ...

Actually, at runtime, ``UserId(42)`` will be just ``42`` not ``int(42)``. So that actually ``b = TcpPacketId(x)``
will work, but ``a = TcpPacketId(127, 0)`` will not work. This is intentional (as discussed in mypy issue),
``TcpPackedId`` works as a cast function, that should be fed by ``PacketId``. Do I need to clarify this in the PEP?

**Comment by gvanrossum** [comment]
I guess you need to clarify this -- using int or str as examples doesn't make this clear at all, since they have the same signature. (And sorry for having such a frazzled memory that I didn't remember this -- but maybe it's been helpful teasing out the ambiguities in the description. :-)

FWIW I think we should take this to python-dev for a formal review period.

**Comment by ilevkivskyi** [comment]
> I guess you need to clarify this -- using int or str as examples doesn't make this clear at all, since they have the same signature.

OK, I will do this and then will post a link to the PR to python-dev.

**Comment by gvanrossum** [comment]
To summarize, the essence here is that to the type checker, UserId is a subclass of int with a constructor that takes an int, while at runtime it is the identity function that returns its argument (hopefully an int :-) unchanged.

**Comment by ilevkivskyi** [comment]
> From this it follows that e.g. adding 1 to a UserId instance is legal and produces an int, not a UserId.

I totally agree with this, there is already this example in the text

**Comment by gvanrossum** [comment]
I just wrote that down to summarize the proposal in the fewest words possible (and because I had never quite realized before that this is how the proposal works out -- it's a very sweet proposal that can be summarized so briefly).

**Comment by gvanrossum** [comment]
The PEP text landed (PR #226). I'd really like to get the typing.py changes in too!

**Comment by gvanrossum** [comment]
NM, the typing.py changes landed too: 807d3a90.
